File: core/recc/packet/role.py
# -*- coding: utf-8 -*-

from typing import Optional, Any, List
from datetime import datetime
from dataclasses import dataclass


@dataclass
class RoleA:
    # Role Table
    slug: str
    name: Optional[str] = None
    description: Optional[str] = None
    extra: Optional[Any] = None
    hidden: Optional[bool] = None
    lock: Optional[bool] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    # RolePermission and Permission Table
    permissions: Optional[List[str]] = None


@dataclass
class CreateRoleQ:
    # Role Table
    slug: str
    name: Optional[str] = None
    description: Optional[str] = None
    extra: Optional[Any] = None
    hidden: Optional[bool] = None
    lock: Optional[bool] = None

    # RolePermission and Permission Table
    permissions: Optional[List[str]] = None

    def normalize_booleans(self) -> None:
        self.hidden = True if self.hidden else False
        self.lock = True if self.lock else False


@dataclass
class UpdateRoleQ:
    # Role Table
    slug: Optional[str] = None
    name: Optional[str] = None
    description: Optional[str] = None
    extra: Optional[Any] = None
    hidden: Optional[bool] = None
    lock: Optional[bool] = None

    # RolePermission and Permission Table
    permissions: Optional[List[str]] = None

File: fe/src/packet/role.ts
export const ROLE_SLUG_OWNER = 'owner';

export interface RoleA {
    slug: string;
    name?: string;
    description?: string;
    extra?: any;
    hidden?: boolean;
    lock?: boolean;
    created_at?: string;
    updated_at?: string;
    permissions?: Array<string>;
}

export interface CreateRoleQ {
    slug: string;
    name?: string;
    description?: string;
    extra?: any;
    hidden?: boolean;
    lock?: boolean;
    permissions?: Array<string>;
}

export interface UpdateRoleQ {
    slug?: string;
    name?: string;
    description?: string;
    extra?: any;
    hidden?: boolean;
    lock?: boolean;
    permissions?: Array<string>;
}
